File: src/commands/evidence/audit.ts
import { execFileSync } from "node:child_process";

import { Flags } from "@oclif/core";

import { QuoinCommand } from "../../base.js";
import { loadMethodCatalog } from "../../advisor/index.js";
import { audit, ratchet } from "../../auditor/index.js";
import {
  baselinePath,
  latestRuns,
  latestScans,
  mockInspectionInput,
  readBaseline,
  readBindings,
} from "../../evidence/index.js";
import {
  checkVersionPremise,
  parseCoverage,
  quireVersion,
  runQuire,
} from "../../quire/index.js";

export default class EvidenceAudit extends QuoinCommand {
  static summary =
    "Audit the evidence store: suspect links, staleness, vacuity.";
  static description = `Reads the store and reports. RUNS NOTHING (ADR-0011 invariant 1) — the
consumer's CI refreshes evidence, and an auditor that could re-run a suite could
also make a finding disappear by re-running it.

Checks:
  suspect-link      the statement changed after the evidence was bound
  stale-evidence    bound to a missing run, or to a run behind HEAD
  vacuous-evidence  every bound symbol was skipped or absent from the run
  undischarged      no evidence is bound at all
  method-conformance   evidence of a kind the declared method does not produce
  unknown-method       the declared method is in no catalog
  insufficient-multiplicity   criticality demands two independent suites
  insufficient-mutation-score criticality demands a mutation score it misses
  unmeasured-mutation-score   a demanded mutation score, with none recorded

--ratchet compares against spec/evidence/baseline.json and fails only on NEW
violations. A gate that fails on the whole existing backlog gets disabled within
a week. Write that baseline with: quoin evidence baseline`;

  static examples = [
    "quoin evidence audit",
    "quoin evidence audit --ratchet --strict",
    "quoin evidence audit --json",
  ];

  static flags = {
    repo: Flags.string({ description: "Repository root.", default: "." }),
    module: Flags.string({
      description: "Module directory supplying the traceability model.",
    }),
    ratchet: Flags.boolean({
      description: "Report only violations absent from the baseline.",
    }),
    strict: Flags.boolean({
      description: "Exit 1 when any reported finding remains.",
    }),
    json: Flags.boolean({ description: "Emit the report as JSON." }),
    "multiplicity-requires": Flags.string({
      description:
        "Criticality values demanding two independent suites. Repeatable. " +
        "Unset by default: no obligation source in the ecosystem declares a " +
        "criticality column, so a built-in default could never fire.",
      multiple: true,
    }),
    "mutation-floor": Flags.string({
      description:
        "Minimum mutation score per criticality, as <criticality>=<0..1>. " +
        "Repeatable. Unset by default, for the same reason as " +
        "--multiplicity-requires: a built-in floor is a rule nobody chose.",
      multiple: true,
    }),
  };

  async run(): Promise<void> {
    const { flags } = await this.parse(EvidenceAudit);

    const premise = checkVersionPremise(quireVersion());
    if (premise) this.error(premise.message, { exit: 2 });

    const args = ["coverage", "--scope", flags.repo, "--json"];
    if (flags.module) args.push("--module", flags.module);
    const parsed = parseCoverage(runQuire(args));
    if (!parsed.ok) this.error(parsed.error.message, { exit: 2 });

    const head = headCommit(flags.repo);
    const mockInspections = mockInspectionInput(flags.repo, head);
    const report = audit({
      obligations: parsed.value.obligations ?? [],
      bindings: readBindings(flags.repo).bindings,
      runs: latestRuns(flags.repo),
      scans: latestScans(flags.repo),
      injections: mockInspections.injections,
      mockInspectionSuites: mockInspections.suites,
      // The SAME module the coverage call above used. Defaulting to the
      // installed roots meant `--module <dir>` derived obligations from one
      // catalog and checked conformance against another — the exact disagreement
      // `src/advisor/methods.ts` opens by warning about (agent-ix/quoin#105).
      catalog: loadMethodCatalog(flags.module ? [flags.module] : undefined),
      headCommit: head,
      // Deliberately unset. No obligation source in the ecosystem declares a
      // `criticality_column` — measured: 2,304 of 2,304 `Acceptance Criteria`
      // tables are `ID | Criteria | Verification` and carry no priority
      // (spec-artifacts-process CR-005) — so a hardcoded `["P0"]` was a rule
      // that could never fire, and would have fired on *everything* the moment
      // a column appeared. `--multiplicity-requires` makes it a choice.
      multiplicityRequires: flags["multiplicity-requires"],
      mutationFloor: parseMutationFloor(flags["mutation-floor"], (message) =>
        this.error(message, { exit: 2 }),
      ),
    });

    const baseline = flags.ratchet ? readBaseline(flags.repo) : null;
    // Whether ratcheting was ACTUALLY applied — `flags.ratchet` is only what
    // was asked for. A missing baseline degrades the run to a full report, and
    // labelling that full report "(new violations only)" told a day-one reader
    // their whole backlog was new violations (#169).
    const ratcheted = baseline !== null;
    const reported = baseline ? ratchet(report, baseline) : report.findings;

    if (flags.ratchet && !ratcheted && !flags.json) {
      this.log(missingBaselineNotice(flags.repo));
    }

    if (flags.json) {
      this.log(
        JSON.stringify(
          {
            findings: reported,
            healthy: report.healthy,
            unevaluated: report.unevaluated,
            ratchet: ratcheted,
          },
          null,
          2,
        ),
      );
    } else if (reported.length === 0 && report.unevaluated.length === 0) {
      this.log(
        `${report.healthy.length} obligation(s) with healthy evidence; nothing to report`,
      );
    } else {
      for (const finding of reported) {
        this.log(`[${finding.severity}] ${finding.kind}: ${finding.summary}`);
      }
      for (const check of report.unevaluated) {
        this.log(
          `[not-evaluated] ${check.check}: ${check.obligation}: ${check.reason}`,
        );
      }
      this.log("");
      this.log(
        `${reported.length} finding(s), ${report.unevaluated.length} check(s) not evaluated, ` +
          `${report.healthy.length} healthy` +
          (ratcheted ? " (new violations only)" : ""),
      );
    }

    if (
      flags.strict &&
      (reported.length > 0 || report.unevaluated.length > 0)
    ) {
      this.exit(1);
    }
  }
}

/**
 * What `--ratchet` prints when there is no baseline to ratchet against.
 *
 * Named path, named remedy: the reader who hits this on day one has not
 * broken anything and has not introduced the backlog they are looking at —
 * they are one `quoin evidence baseline` away from the gate they asked for.
 */
export function missingBaselineNotice(repo: string): string {
  return (
    `--ratchet requested but ${baselinePath(repo)} does not exist; ` +
    `reporting the full backlog, not new violations. ` +
    `Write the baseline with: quoin evidence baseline`
  );
}

/**
 * The newest recorded run per suite, by `timestamp`.
 *
 * This used to take `listRuns(...).at(-1)` — the lexicographically last
 * `<commit12>.json`, which is uniformly random hex and so picked the newest run
 * with probability 1/n. It drove `stale-evidence` (`run.commit !== headCommit`)
 * and `vacuous-evidence` (which symbols the run reported), so a fresh run at
 * HEAD could still be reported stale, and re-recording could not fix it
 * (agent-ix/quoin#104).
 */
function headCommit(repo: string): string | undefined {
  try {
    return execFileSync("git", ["-C", repo, "rev-parse", "HEAD"], {
      encoding: "utf8",
    }).trim();
  } catch {
    return undefined;
  }
}

/**
 * `--mutation-floor P0=0.8` → `{ P0: 0.8 }`.
 *
 * Rejected rather than ignored: a malformed entry means the operator asked for
 * a threshold and would otherwise get a clean report saying nothing was below
 * it. A floor that silently does not apply is worse than no floor, because it
 * reads as a passing gate.
 */
export function parseMutationFloor(
  entries: string[] | undefined,
  fail: (message: string) => never,
): Record<string, number> | undefined {
  if (!entries || entries.length === 0) return undefined;
  const out: Record<string, number> = {};
  for (const entry of entries) {
    const [criticality, raw] = entry.split("=");
    const score = Number(raw);
    if (!criticality || raw === undefined || Number.isNaN(score)) {
      fail(
        `--mutation-floor expects <criticality>=<score>, got '${entry}'. ` +
          `Example: --mutation-floor P0=0.8`,
      );
    }
    if (score < 0 || score > 1) {
      // A ratio, not a percentage. `--mutation-floor P0=80` would otherwise be
      // a floor nothing can ever reach, reported as every P0 failing.
      fail(
        `--mutation-floor score must be a ratio in [0, 1], got '${raw}'. ` +
          `80% is 0.8, not 80.`,
      );
    }
    out[criticality] = score;
  }
  return out;
}

File: rust/crates/quoin-cli/src/evidence/audit.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Agent-IX

//! The retained read-only `quoin evidence audit` adapter.

use clap::{Arg, ArgAction, ArgMatches, Command};
use quoin_core::protocol::{Outcome, Response};

use super::{json_arg, repo_arg, required, revision};
use crate::core_runtime::invoke;

pub(super) fn command() -> Command {
    Command::new("audit")
        .about("Audit the evidence store: suspect links, staleness, vacuity")
        .arg(repo_arg())
        .arg(Arg::new("module").long("module").action(ArgAction::Append))
        .arg(
            Arg::new("ratchet")
                .long("ratchet")
                .action(ArgAction::SetTrue),
        )
        .arg(Arg::new("strict").long("strict").action(ArgAction::SetTrue))
        .arg(json_arg())
        .arg(
            Arg::new("multiplicity-requires")
                .long("multiplicity-requires")
                .action(ArgAction::Append),
        )
        .arg(
            Arg::new("mutation-floor")
                .long("mutation-floor")
                .action(ArgAction::Append),
        )
        .arg(Arg::new("independence-policy").long("independence-policy"))
}

pub(super) fn run(arguments: &ArgMatches) -> Result<Response, String> {
    let repo = required(arguments, "repo")?;
    let modules = values(arguments, "module");
    let coverage = invoke(
        "quire.coverage",
        &serde_json::json!({ "scope": repo, "modules": modules }),
    )?;
    if !coverage.outcome.carries_payload() {
        return Ok(coverage);
    }
    let obligations = field(&coverage.payload, "obligations")?;
    let policy = policy(arguments, &obligations)?;
    let head = revision(&repo);
    let store = invoke(
        "evidence.audit_inputs",
        &serde_json::json!({ "repo": repo, "head_commit": if head.is_empty() { None } else { Some(head.as_str()) }, "independence_policy": policy }),
    )?;
    if !store.outcome.carries_payload() {
        return Ok(store);
    }
    let (accepted, missing_baseline) = baseline(arguments, &repo)?;
    let catalog_request = if modules.is_empty() {
        serde_json::json!({})
    } else {
        serde_json::json!({ "roots": modules })
    };
    let catalog = invoke("catalog.methods", &catalog_request)?;
    if !catalog.outcome.carries_payload() {
        return Ok(catalog);
    }
    let input = serde_json::json!({
        "obligations": obligations,
        "bindings": field(&store.payload, "bindings")?,
        "runs": field(&store.payload, "runs")?,
        "scans": field(&store.payload, "scans")?,
        "injections": field(&store.payload, "injections")?,
        "mockInspectionSuites": field(&store.payload, "mock_inspection_suites")?,
        "vacuousScanSuites": field(&store.payload, "vacuous_scan_suites")?,
        "independence": field(&store.payload, "independence")?,
        "catalog": catalog.payload,
        "headCommit": if head.is_empty() { None } else { Some(head.as_str()) },
        "multiplicityRequires": values(arguments, "multiplicity-requires"),
        "mutationFloor": mutation_floor(&values(arguments, "mutation-floor"))?,
        "independencePolicy": policy,
    });
    let request = serde_json::json!({ "input": input, "accepted": accepted });
    let audited = invoke("auditor.audit", &request)?;
    if !audited.outcome.carries_payload() {
        return Ok(audited);
    }
    let report = audited
        .payload
        .get("report")
        .ok_or_else(|| "auditor.audit did not return report".to_owned())?;
    let reported = audited
        .payload
        .get("reported")
        .or_else(|| report.get("findings"))
        .ok_or_else(|| "auditor.audit did not return findings".to_owned())?;
    let unevaluated = report
        .get("unevaluated")
        .and_then(serde_json::Value::as_array)
        .map_or(0, Vec::len);
    let findings = reported.as_array().map_or(0, Vec::len);
    let rendered = if arguments.get_flag("json") {
        serde_json::to_string_pretty(&serde_json::json!({
            "findings": reported,
            "healthy": report.get("healthy"),
            "unevaluated": report.get("unevaluated"),
            "ratchet": audited.payload.get("reported").is_some(),
            "independence": audited.payload.get("independence"),
        }))
        .map_err(|error| error.to_string())?
    } else {
        render_text(
            report,
            reported,
            audited.payload.get("reported").is_some(),
            missing_baseline,
        )?
    };
    Ok(Response {
        payload: serde_json::json!({ "rendered": rendered }),
        diagnostics: audited.diagnostics,
        outcome: if arguments.get_flag("strict") && (findings > 0 || unevaluated > 0) {
            Outcome::Partial
        } else {
            audited.outcome
        },
    })
}

fn policy(
    arguments: &ArgMatches,
    obligations: &serde_json::Value,
) -> Result<serde_json::Value, String> {
    let Some(path) = arguments.get_one::<String>("independence-policy") else {
        return Ok(serde_json::Value::Null);
    };
    let text =
        std::fs::read_to_string(path).map_err(|error| format!("cannot read {path}: {error}"))?;
    let known_obligations = obligations
        .as_array()
        .into_iter()
        .flatten()
        .filter_map(|row| row.get("id").and_then(serde_json::Value::as_str))
        .collect::<Vec<_>>();
    let parsed = invoke(
        "evidence.parse_policy",
        &serde_json::json!({ "text": text, "known_obligations": known_obligations }),
    )?;
    if !parsed.outcome.carries_payload() {
        return Err("independence policy was refused".to_owned());
    }
    field(&parsed.payload, "policy")
}

fn baseline(
    arguments: &ArgMatches,
    repo: &str,
) -> Result<(Option<serde_json::Value>, bool), String> {
    if !arguments.get_flag("ratchet") {
        return Ok((None, false));
    }
    let response = invoke(
        "evidence.read_baseline",
        &serde_json::json!({ "repo": repo }),
    )?;
    if !response.outcome.carries_payload() {
        return Ok((None, false));
    }
    let Some(file) = response
        .payload
        .get("baseline")
        .filter(|value| !value.is_null())
    else {
        return Ok((None, true));
    };
    Ok((Some(field(file, "accepted")?), false))
}

fn mutation_floor(entries: &[String]) -> Result<serde_json::Value, String> {
    if entries.is_empty() {
        return Ok(serde_json::Value::Null);
    }
    let mut floor = serde_json::Map::new();
    for entry in entries {
        let Some((criticality, raw)) = entry.split_once('=') else {
            return Err(format!(
                "--mutation-floor expects <criticality>=<score>, got '{entry}'. Example: --mutation-floor P0=0.8"
            ));
        };
        let score = raw.parse::<f64>().map_err(|_| format!("--mutation-floor expects <criticality>=<score>, got '{entry}'. Example: --mutation-floor P0=0.8"))?;
        if criticality.is_empty() || !(0.0..=1.0).contains(&score) {
            return Err(format!(
                "--mutation-floor score must be a ratio in [0, 1], got '{raw}'. 80% is 0.8, not 80."
            ));
        }
        floor.insert(criticality.to_owned(), serde_json::json!(score));
    }
    Ok(serde_json::Value::Object(floor))
}

fn values(arguments: &ArgMatches, name: &str) -> Vec<String> {
    arguments
        .get_many::<String>(name)
        .map_or_else(Vec::new, |values| values.cloned().collect())
}

fn field(value: &serde_json::Value, name: &str) -> Result<serde_json::Value, String> {
    value
        .get(name)
        .cloned()
        .ok_or_else(|| format!("core response did not return {name}"))
}

fn render_text(
    report: &serde_json::Value,
    reported: &serde_json::Value,
    ratcheted: bool,
    missing_baseline: bool,
) -> Result<String, String> {
    let mut lines = Vec::new();
    if missing_baseline {
        lines.push("--ratchet requested but no baseline exists; reporting the full backlog, not new violations. Write the baseline with: quoin evidence baseline".to_owned());
    }
    let findings = reported
        .as_array()
        .ok_or_else(|| "auditor.audit did not return findings".to_owned())?;
    let unevaluated = report
        .get("unevaluated")
        .and_then(serde_json::Value::as_array)
        .ok_or_else(|| "auditor.audit did not return unevaluated checks".to_owned())?;
    if findings.is_empty() && unevaluated.is_empty() {
        lines.push(format!(
            "{} obligation(s) with healthy evidence; nothing to report",
            report
                .get("healthy")
                .and_then(serde_json::Value::as_array)
                .map_or(0, Vec::len)
        ));
    } else {
        for finding in findings {
            lines.push(format!(
                "[{}] {}: {}",
                finding
                    .get("severity")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
                finding
                    .get("kind")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
                finding
                    .get("summary")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or("")
            ));
        }
        for check in unevaluated {
            lines.push(format!(
                "[not-evaluated] {}: {}: {}",
                check
                    .get("check")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
                check
                    .get("obligation")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or(""),
                check
                    .get("reason")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or("")
            ));
        }
        lines.push(String::new());
        lines.push(format!(
            "{} finding(s), {} check(s) not evaluated, {} healthy{}",
            findings.len(),
            unevaluated.len(),
            report
                .get("healthy")
                .and_then(serde_json::Value::as_array)
                .map_or(0, Vec::len),
            if ratcheted {
                " (new violations only)"
            } else {
                ""
            }
        ));
    }
    Ok(lines.join("\n"))
}
